File: zebar/caelestia/media-art.js
// Album art for the currently playing track.
//
// Zebar's media provider carries title/artist/album/position and NO artwork --
// grepping the vendored bundle for thumbnail/artwork/albumArt/coverArt returns
// nothing, so there is no field being overlooked. Direct user report: "it
// still can't fetch the current playing media image (this was possible in my
// previous yasb bar)". Correct on both counts; yasb read the Windows SMTC API
// directly rather than going through a provider, and tools/media-art.cs now
// does the same for this pack.
//
// Same split as net-stats.js and gpu-stats.js: pure parsing here, testable
// without a shell, and a stateful fetcher that owns the caching and the
// process discipline.

export const MEDIA_ART_PATH =
  'C:\\Users\\PC\\Documents\\git\\setup\\zebar\\caelestia\\tools\\media-art.exe';

export function artCommand() {
  return { program: MEDIA_ART_PATH, args: [] };
}

// One key for "which track is this". Case- and whitespace-insensitive,
// because players are not consistent about either between the metadata they
// report to the provider and the metadata they report to SMTC -- and a key
// that disagrees between the two would re-fetch the same cover forever.
export function trackKey(title, artist) {
  const norm = (s) => String(s ?? '').trim().replace(/\s+/g, ' ').toLowerCase();
  return `${norm(title)}\u0000${norm(artist)}`;
}

/**
 * Parses media-art.exe's single output line: `<title>\t<artist>\t<base64>`.
 *
 * Never throws; returns null for anything malformed. The tool prints nothing
 * at all on every failure it knows about (no session, no thumbnail, timeout),
 * so an empty string is the ORDINARY case, not an error -- plenty of tracks
 * genuinely have no cover.
 */
export function parseArt(stdout) {
  if (typeof stdout !== 'string') return null;
  const line = stdout.replace(/\r?\n$/, '');
  if (!line) return null;

  const tab = line.indexOf('\t');
  if (tab === -1) return null;
  const tab2 = line.indexOf('\t', tab + 1);
  if (tab2 === -1) return null;

  const title = line.slice(0, tab);
  const artist = line.slice(tab + 1, tab2);
  const b64 = line.slice(tab2 + 1);
  if (!b64) return null;
  // Cheap sanity check: base64 only. Catches a tool that started printing a
  // diagnostic before anyone notices it rendering as a broken image.
  if (!/^[A-Za-z0-9+/]+={0,2}$/.test(b64)) return null;

  // image/png is a lie only in the MIME label: SMTC thumbnails are usually
  // JPEG. Browsers sniff the actual bytes for data: URIs, so the label does
  // not have to be right -- but PNG is the label app-icon.js already uses, so
  // both stay consistent.
  return { title, artist, dataUrl: `data:image/png;base64,${b64}` };
}

/**
 * Fetches artwork for a track, at most once per track.
 *
 * Three things this has to get right, all of which cost something if wrong:
 *
 * 1. **One spawn per track, not one per render.** The panel re-renders on
 *    every provider emission -- once a second or faster. Spawning a process
 *    that often is exactly the pattern that has twice orphaned a helper onto
 *    zebar's listening socket and blanked every widget.
 * 2. **Discard a stale answer.** The fetch is async; the song can change
 *    while it is in flight. media-art.exe echoes back the track its image is
 *    actually for, which is the only reliable way to tell -- so an answer for
 *    a track that is no longer current is dropped rather than shown.
 * 3. **Remember failures too.** A track with no cover must be asked once, not
 *    once per second forever.
 */
export function createMediaArt(shell, { maxEntries = 24 } = {}) {
  const cache = new Map();   // trackKey -> dataUrl | null (null = asked, none)
  let inFlight = false;

  return {
    /** Synchronous cache read: what the renderer calls on every frame. */
    cached(title, artist) {
      const key = trackKey(title, artist);
      return cache.has(key) ? cache.get(key) : undefined;
    },

    /**
     * Fetches if this track has not been asked about yet. Resolves to the
     * data URL, or null when there is no art (or no way to get it).
     */
    async fetch(title, artist) {
      const key = trackKey(title, artist);
      if (cache.has(key)) return cache.get(key);
      if (inFlight) return undefined;          // try again on the next render
      if (!shell || typeof shell.shellExec !== 'function') return null;

      inFlight = true;
      try {
        const cmd = artCommand();
        let stdout = '';
        try {
          const res = await shell.shellExec(cmd.program, cmd.args);
          stdout = res && typeof res.stdout === 'string' ? res.stdout : '';
        } catch (e) {
          // Nonzero exit is the tool's documented "no art" signal, not a
          // fault worth logging on every silent track.
          stdout = '';
        }

        const parsed = parseArt(stdout);
        if (!parsed) {
          remember(cache, key, null, maxEntries);
          return null;
        }

        // (2): store the answer under the track it is ACTUALLY for, which may
        // not be the one that was asked about.
        const actualKey = trackKey(parsed.title, parsed.artist);
        remember(cache, actualKey, parsed.dataUrl, maxEntries);
        if (actualKey !== key) {
          // The song changed mid-fetch. Do not attribute this cover to the
          // track that was asked about; leave that one unasked so the next
          // render retries it.
          return undefined;
        }
        return parsed.dataUrl;
      } finally {
        inFlight = false;
      }
    },
  };
}

// A plain bounded LRU-ish map: oldest insertion evicted first. A playlist can
// run for hours, and each entry is a base64 image worth ~100KB.
function remember(cache, key, value, maxEntries) {
  if (cache.has(key)) cache.delete(key);
  cache.set(key, value);
  while (cache.size > maxEntries) {
    const oldest = cache.keys().next().value;
    cache.delete(oldest);
  }
}

File: zebar/caelestia/tools/media-art.cs
// Rebuild after editing:
//   C:\Windows\Microsoft.NET\Framework64\v4.0.30319\csc.exe /nologo /target:exe \
//     /optimize+ /out:media-art.exe \
//     /r:"C:\Program Files (x86)\Windows Kits\10\UnionMetadata\10.0.22621.0\Windows.winmd" \
//     /r:"C:\Windows\Microsoft.NET\Framework64\v4.0.30319\System.Runtime.WindowsRuntime.dll" \
//     media-art.cs
//
// Prints a base64-encoded image of the CURRENTLY PLAYING track's album art
// to stdout, plus the track it belongs to, as one line:
//
//   <title>\t<artist>\t<base64>
//
// Prints nothing and exits non-zero on any failure -- no session, no
// thumbnail, a player that does not publish one, a timeout. dashboard.js
// treats "no output" as "keep the placeholder glyph", the same fail-soft
// contract app-icon.exe and fullscreen-detect.exe already have.
//
// WHY THIS TOOL EXISTS. Direct user report: "it still can't fetch the current
// playing media image (this was possible in my previous yasb bar)". They are
// right on both counts. Zebar's media provider carries title/artist/album/
// position and nothing else -- grepping the vendored bundle for
// thumbnail/artwork/albumArt/coverArt returns zero hits, so there is no field
// being missed. yasb could do it because it read the same Windows API this
// tool reads, directly, rather than going through a provider.
//
// The API is SMTC -- GlobalSystemMediaTransportControlsSessionManager, the
// thing behind the volume-flyout media popup. It is the only universal source
// on Windows: it works for Spotify, browsers, foobar2000 and anything else
// that registers a transport session, and it is what the OS itself displays.
//
// TITLE AND ARTIST ARE ECHOED BACK ON PURPOSE. Fetching art costs a process
// spawn and a WinRT round trip, so the caller caches per track -- but a cache
// keyed on the caller's idea of "current track" races the player: the art that
// arrives may already belong to the previous song. Returning the track this
// image is actually FOR lets the caller discard a stale answer instead of
// showing the wrong cover. (The same race exists in app-icon.exe, where it is
// harmless because a window's icon does not change under it; here it is not.)
//
// A note on the toolchain: csc.exe here is C# 5, which is fine -- async/await
// is C# 5's own headline feature, and .AsTask() comes from
// System.Runtime.WindowsRuntime. What is NOT available: interpolated strings,
// discards (out _), and expression-bodied members. Keep to C# 5.

using System;
using System.IO;
using System.Runtime.InteropServices.WindowsRuntime;
using System.Threading.Tasks;
using Windows.Media.Control;
using Windows.Storage.Streams;

static class MediaArt
{
    // Every await is raced against this. A hung player (or a WinRT call that
    // never returns because the session died mid-request) must not leave a
    // process parked forever: this pack has twice been broken by a helper
    // outliving its parent zebar and inheriting zebar's listening socket on
    // port 6124, after which every later start paints nothing at all. A tool
    // that always exits cannot become that orphan.
    const int TimeoutMs = 4000;

    static int Main()
    {
        try
        {
            Task<int> work = Run();
            if (!work.Wait(TimeoutMs)) return 3;
            return work.Result;
        }
        catch (Exception)
        {
            // Deliberately silent. Any diagnostic on stdout would be parsed as
            // an answer by the caller; stderr is not read by anyone.
            return 1;
        }
    }

    static async Task<int> Run()
    {
        GlobalSystemMediaTransportControlsSessionManager manager =
            await GlobalSystemMediaTransportControlsSessionManager.RequestAsync().AsTask();
        if (manager == null) return 2;

        GlobalSystemMediaTransportControlsSession session = manager.GetCurrentSession();
        if (session == null) return 2;

        GlobalSystemMediaTransportControlsSessionMediaProperties props =
            await session.TryGetMediaPropertiesAsync().AsTask();
        if (props == null) return 2;

        IRandomAccessStreamReference thumbRef = props.Thumbnail;
        // A session with no artwork is ordinary, not an error: a browser tab
        // playing a bare audio file has none.
        if (thumbRef == null) return 2;

        byte[] bytes;
        using (IRandomAccessStreamWithContentType stream = await thumbRef.OpenReadAsync().AsTask())
        {
            if (stream == null || stream.Size == 0) return 2;
            // Guard against a pathological stream: the caller turns this into
            // a data: URI, and a multi-megabyte one would bloat the DOM for a
            // 240px tile. Real SMTC thumbnails are tens of KB.
            if (stream.Size > 8 * 1024 * 1024) return 2;

            uint size = (uint)stream.Size;
            Windows.Storage.Streams.Buffer buffer = new Windows.Storage.Streams.Buffer(size);
            IBuffer read = await stream.ReadAsync(buffer, size, InputStreamOptions.None).AsTask();
            bytes = read.ToArray();
        }

        if (bytes == null || bytes.Length == 0) return 2;

        string title = props.Title == null ? "" : props.Title;
        string artist = props.Artist == null ? "" : props.Artist;
        // Tabs separate the three fields, so strip any that appear in the
        // metadata itself rather than emitting a line the caller cannot split.
        title = title.Replace('\t', ' ').Replace('\r', ' ').Replace('\n', ' ');
        artist = artist.Replace('\t', ' ').Replace('\r', ' ').Replace('\n', ' ');

        Console.Out.Write(title);
        Console.Out.Write('\t');
        Console.Out.Write(artist);
        Console.Out.Write('\t');
        Console.Out.Write(Convert.ToBase64String(bytes));
        Console.Out.Flush();
        return 0;
    }
}
